sklearn_numba_dpex/kmeans/drivers.py

```diff
@@ -426,17 +412,34 @@ def _lloyd(
                 reduce_centroid_shifts_kernel(centroid_shifts)
             )
 
-            pseudo_inertia, *_ = dpctl.tensor.asnumpy(
-                reduce_inertia_kernel(per_sample_pseudo_inertia)
-            )
-
-            # TODO: should we drop this check ?
-            # (it should not be needed because we have theoritical guarantee that
-            # inertia decreases, if it doesn't it should only be because of rounding
-            # errors)
-            if pseudo_inertia < best_pseudo_inertia:
-                best_pseudo_inertia = pseudo_inertia
-                copyto_kernel(centroids_t, best_centroids_t)
+            if verbose:
+                # ???: verbosity comes at the cost of performance since it triggers
+                # computing exact inertia at each iteration. Shouldn't this be
+                # documented ?
+                inertia, *_ = dpctl.tensor.asnumpy(
+                    reduce_inertia_kernel(per_sample_inertia)
+                )
+                print(f"Iteration {n_iteration}, inertia {inertia}")
+
+            # ???: unlike sklearn, sklearn_intelex checks that pseudo_inertia decreases
+            # and keep an additional copy of centroids that is updated only if the
+            # value of the pseudo_inertia is smaller than all the past values.
+            #
+            # (the check should not be needed because we have theoritical guarantee
+            # that inertia decreases, if it doesn't it should only be because of
+            # rounding errors ?)
+            #
+            # To this purpose this code could be inserted here:
+            #
+            # if pseudo_inertia < best_pseudo_inertia:
+            #     best_pseudo_inertia = pseudo_inertia
+            #     copyto_kernel(centroids_t, best_centroids_t)
+            #
+            # Note that what that is saved as "best centroid" is the array before the
+            # update, to which the pseudo inertia that is computed at this iteration
+            # refers. For this reason, this strategy is not compatible with sklearn
+            # unit tests, that consider new_centroids_t (the array after the update)
+            # to be the best centroids at  each iteration.
```

Comment:
nit:

```suggestion
            # to be the best centroids at each iteration.
```

Reply:
BTW, shall we open a PR to implement this "best centroid" strategy in scikit-learn's Cython k-means? It seems more correct from a mathematical point of view but I am not 100% sure why an update would increase the inertia. Maybe it's just rounding errors, in which case the current code might be fine.

This best centroid check could be useful if we were to implement early stopping on a validation set though.

Maybe @jeremiedbb has an opinion on this.

Reply:
I don't have an opinion on this, frankly I doubt it really has a practical implication and sounds like over-optimization, ideally someone with more knowledge about KMeans could help us decide.

In any case, transitioning to either one or the other strategy is straightforward, and we will stick to whatever strategy conforms to sklearn unit tests.

Maybe the wall of inline text I let in the code gives too much importance to the issue :grin:

Reply:
I think it's fine to keep the inline comment.

Reply:
Inertia is guaranteed to decrease at each iteration, even when a center is relocated. The only way I could see it increase is due to rounding errors, but then it would only increase by a tiny amount. I can't see the benefit of this "best centroids" strategy.